ledger/participant-state/kvutils/tools/src/test/scala/com/daml/ledger/participant/state/kvutils/tools/IntegrityCheckerSpec.scala

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.tools
+
+import com.daml.ledger.participant.state.kvutils.export.FileBasedLedgerDataExporter.WriteSet
+import com.daml.ledger.participant.state.kvutils.tools.export.{
+  CommitStrategySupport,
+  IntegrityChecker
+}
+import com.daml.ledger.validator.LedgerStateOperations.{Key, Value}
+import com.google.protobuf.ByteString
+import org.mockito.ArgumentMatchers._
+import org.mockito.Mockito._
+import org.scalatest.mockito.MockitoSugar
+import org.scalatest.{Matchers, WordSpec}
+
+class IntegrityCheckerSpec extends WordSpec with Matchers with MockitoSugar {
+  "compareSameSizeWriteSets" should {
+    "return None in case strategy cannot explain difference" in {
+      val mockCommitStrategySupport = mock[CommitStrategySupport[Unit]]
+      when(mockCommitStrategySupport.explainMismatchingValue(any[Key], any[Value], any[Value]))
+        .thenReturn(None)
+      val instance = new IntegrityChecker[Unit](mockCommitStrategySupport)
+
+      instance.compareSameSizeWriteSets(toWriteSet("key" -> "a"), toWriteSet("key" -> "b")) shouldBe None
+    }
+
+    "return None in case of no difference" in {
+      val mockCommitStrategySupport = mock[CommitStrategySupport[Unit]]
+      val instance = new IntegrityChecker[Unit](mockCommitStrategySupport)
+      val aWriteSet = toWriteSet("key" -> "value")
+
+      instance.compareSameSizeWriteSets(aWriteSet, aWriteSet) shouldBe None
+    }
+
+    "return explanation from strategy in case it can explain the difference" in {
+      val mockCommitStrategySupport = mock[CommitStrategySupport[Unit]]
+      when(mockCommitStrategySupport.explainMismatchingValue(any[Key], any[Value], any[Value]))
+        .thenReturn(Some("expected explanation"))
+      val instance = new IntegrityChecker[Unit](mockCommitStrategySupport)
+
+      val actual =
+        instance.compareSameSizeWriteSets(toWriteSet("key" -> "a"), toWriteSet("key" -> "b"))
+
+      actual match {
+        case Some(explanation) => explanation should include("expected explanation")
```

Comment:
Can we match the full string in this case? I'd like to make sure there's no instance of "List(".

Reply:
We could, however, I didn't want to replicate the full code for generating the explanation string. Instead, here we just count the number of new lines in the result which will ensure that we have converted list elements into separate lines.